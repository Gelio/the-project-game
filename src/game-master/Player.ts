import { Point } from '../common/Point';
import { TeamId } from '../common/TeamId';
import { Piece } from './models/Piece';

export class Player {
  public isConnected: boolean;
  public teamId: TeamId;
  public isLeader: boolean;
  public isBusy = false;

  public playerId: number;
  public position: Point | null = null;
  public heldPiece: Piece | null = null;
}
