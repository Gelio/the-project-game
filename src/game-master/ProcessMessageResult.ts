import { MessageWithRecipient } from '../interfaces/MessageWithRecipient';

export interface ValidMessageResult<T> {
  valid: true;
  delay: number;
  responseMessage: Promise<MessageWithRecipient<T>>;
}

export interface InvalidMessageResult<T> {
  valid: false;
  reason: string;
}

export type ProcessMessageResult<T> = ValidMessageResult<T> | InvalidMessageResult<T>;
