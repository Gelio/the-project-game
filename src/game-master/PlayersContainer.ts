import { TeamId } from '../common/TeamId';
import { Player } from './Player';

export class PlayersContainer {
  public readonly players: Player[] = [];

  public addPlayer(player: Player) {
    if (this.players.indexOf(player) !== -1) {
      throw new Error('Player already added');
    }
    this.players.push(player);
  }

  public removePlayer(player: Player) {
    const playerIndex = this.players.indexOf(player);
    if (playerIndex === -1) {
      throw new Error('Player is not added');
    }
    this.players.splice(playerIndex, 1);
  }

  public getPlayersFromTeam(teamId: TeamId) {
    return this.players.filter(player => player.teamId === teamId);
  }

  public getConnectedPlayers() {
    return this.players.filter(player => player.isConnected);
  }

  public getPlayerById(playerId: number) {
    return this.players.find(player => player.playerId === playerId);
  }
}
