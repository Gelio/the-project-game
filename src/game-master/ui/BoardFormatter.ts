import { Board } from '../models/Board';
import { Piece } from '../models/Piece';
import { NeutralAreaTile } from '../models/tiles/NeutralAreaTile';
import { TeamAreaTile } from '../models/tiles/TeamAreaTile';
import { Player } from '../Player';

export class BoardFormatter {
  public displayTile(board: Board, tile: NeutralAreaTile | TeamAreaTile): string {
    if (tile.player) {
      return this.displayPlayer(tile.player);
    }

    if (tile.piece) {
      return this.displayPiece(tile.piece);
    }

    if (tile instanceof NeutralAreaTile) {
      return ' ';
    } else if (tile instanceof TeamAreaTile) {
      return this.displayTeamAreaTile(board, tile);
    }

    throw new Error('Unknown tile');
  }

  private displayPlayer(player: Player): string {
    if (player.teamId === 1) {
      return '{blue-fg}{white-bg}P{/}';
    } else {
      return '{red-fg}{white-bg}P{/}';
    }
  }

  private displayPiece(piece: Piece): string {
    if (piece.isSham) {
      return '{grey-fg}X{/}';
    }

    return '{yellow-fg}X{/}';
  }

  private displayTeamAreaTile(board: Board, tile: TeamAreaTile): string {
    const backgroundColor = tile.y < board.size.taskArea ? '{blue-bg}' : '{red-bg}';

    if (tile.hasGoal) {
      if (tile.hasCompletedGoal) {
        return `${backgroundColor}{yellow-fg}G{/}`;
      }

      return `${backgroundColor}{white-fg}G{/}`;
    }

    return `${backgroundColor} {/}`;
  }
}
