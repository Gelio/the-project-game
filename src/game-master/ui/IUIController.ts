import { Board } from '../models/Board';
import { Scoreboard } from '../models/Scoreboard';

import { GameMasterOptions } from '../GameMaster';
import { PlayersContainer } from '../PlayersContainer';

export interface UIController {
  init(): void;
  destroy(): void;
  render(): void;
  log(lines: string | string[]): void;
  updateBoard(board: Board): void;
  updateGameInfo(
    currentRound: number,
    gameMasterOptions: GameMasterOptions,
    board: Board,
    scoreboard: Scoreboard,
    playersContainer: PlayersContainer
  ): void;
}
