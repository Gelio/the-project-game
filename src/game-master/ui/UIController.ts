import * as blessed from 'blessed';

import { Service } from '../../interfaces/Service';

import { config } from '../config';
import { GameMasterOptions } from '../GameMaster';
import { PlayersContainer } from '../PlayersContainer';

import { Board } from '../models/Board';
import { Scoreboard } from '../models/Scoreboard';

import { BoardFormatter } from './BoardFormatter';

export type BoxFactoryFn = (options: blessed.Widgets.BoxOptions) => blessed.Widgets.BoxElement;

export class UIController implements Service {
  private readonly screen: blessed.Widgets.Screen;
  private readonly boardFormatter: BoardFormatter;
  private readonly boxFactoryFn: BoxFactoryFn;

  private boardBox: blessed.Widgets.BoxElement;
  private infoBox: blessed.Widgets.BoxElement;
  private logsBox: blessed.Widgets.BoxElement;

  constructor(screen: blessed.Widgets.Screen, boxFactoryFn: BoxFactoryFn) {
    this.screen = screen;
    this.boardFormatter = new BoardFormatter();
    this.boxFactoryFn = boxFactoryFn;
  }

  public init() {
    this.boardBox = this.boxFactoryFn({
      width: '50%',
      height: '100%',
      border: {
        type: 'line'
      },
      tags: true
    });
    this.boardBox.setContent(`${config.uiLabelStyle}Board{/}`);

    this.infoBox = this.boxFactoryFn({
      left: '50%',
      width: '50%',
      height: '30%',
      border: {
        type: 'line'
      },
      tags: true
    });
    this.infoBox.setContent(`${config.uiLabelStyle}Info{/}`);

    this.logsBox = this.boxFactoryFn({
      left: '50%',
      top: '30%',
      width: '50%',
      height: '70%',
      border: {
        type: 'line'
      },
      tags: true,
      scrollable: true
    });
    this.logsBox.setContent(`${config.uiLabelStyle}Logs{/}`);

    this.screen.append(this.boardBox);
    this.screen.append(this.infoBox);
    this.screen.append(this.logsBox);
    this.render();
  }

  public destroy() {
    this.screen.destroy();
  }

  public render() {
    this.screen.render();
  }

  public log(lines: string | string[]) {
    this.logsBox.pushLine(lines);
    this.logsBox.setScrollPerc(100);
    this.render();
  }

  public updateBoard(board: Board) {
    this.boardBox.setContent(`${config.uiLabelStyle}Board{/}`);
    const tiles = board.tiles;

    for (let y = 0; y < tiles[0].length; y++) {
      const line = tiles.map(tilesRow => this.boardFormatter.displayTile(board.size, tilesRow[y]));
      this.boardBox.pushLine(line.join(''));
    }

    // FIXME: check if `render` is needed and when should it be called
    this.render();
  }

  public updateGameInfo(
    currentRound: number,
    gameMasterOptions: GameMasterOptions,
    board: Board,
    scoreboard: Scoreboard,
    playersContainer: PlayersContainer
  ) {
    // TODO: add tests

    this.infoBox.setContent(`${config.uiLabelStyle}Info{/}`);

    this.infoBox.pushLine(`Round: ${currentRound} / ${gameMasterOptions.gamesLimit}`);
    this.infoBox.pushLine(
      `{blue-fg}Blue{/} team score: ${scoreboard.team1Score} / ${scoreboard.scoreLimit}`
    );
    this.infoBox.pushLine(
      `{red-fg}Red{/} team score: ${scoreboard.team2Score} / ${scoreboard.scoreLimit}`
    );

    const blueTeamPlayers = playersContainer.getPlayersFromTeam(1).length;
    const blueTeamCapacity = gameMasterOptions.teamSizes[1];
    this.infoBox.pushLine('');
    this.infoBox.pushLine(
      `{blue-fg}Blue{/} team players: ${blueTeamPlayers} / ${blueTeamCapacity}`
    );
    const redTeamPlayers = playersContainer.getPlayersFromTeam(2).length;
    const redTeamCapacity = gameMasterOptions.teamSizes[2];
    this.infoBox.pushLine(`{red-fg}Red{/} team players: ${redTeamPlayers} / ${redTeamCapacity}`);

    this.infoBox.pushLine('');
    const pickedUpPiecesCount = board.pieces.filter(piece => piece.isPickedUp).length;
    const shamCount = board.pieces.filter(piece => piece.isSham).length;
    this.infoBox.pushLine(
      `Pieces: ${board.pieces.length} (${pickedUpPiecesCount} picked up, ${shamCount} shams)`
    );
  }
}
