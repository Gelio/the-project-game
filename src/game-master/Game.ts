import { LoggerInstance } from 'winston';

import { Player } from './Player';
import { PlayersContainer } from './PlayersContainer';

import { GameDefinition } from '../interfaces/GameDefinition';
import { Message } from '../interfaces/Message';

import { ActionInvalidMessage } from '../interfaces/messages/ActionInvalidMessage';
import { ActionValidMessage } from '../interfaces/messages/ActionValidMessage';
import {
  GameFinishedMessage,
  GameFinishedMessagePayload
} from '../interfaces/messages/GameFinishedMessage';
import { GameStartedMessage } from '../interfaces/messages/GameStartedMessage';
import { PlayerDisconnectedMessage } from '../interfaces/messages/PlayerDisconnectedMessage';
import { PlayerHelloMessage } from '../interfaces/messages/PlayerHelloMessage';

import { RegisterGameRequest } from '../interfaces/requests/RegisterGameRequest';
import { UnregisterGameRequest } from '../interfaces/requests/UnregisterGameRequest';
import { RegisterGameResponse } from '../interfaces/responses/RegisterGameResponse';
import { UnregisterGameResponse } from '../interfaces/responses/UnregisterGameResponse';

import { Board } from './models/Board';
import { Scoreboard } from './models/Scoreboard';

import { GameState } from './GameState';
import { ProcessMessageResult } from './ProcessMessageResult';

import { UIController } from './ui/IUIController';

import { PlayerMessageHandler } from './game-logic/PlayerMessageHandler';

import { Communicator } from '../common/Communicator';
import { COMMUNICATION_SERVER_ID, GAME_MASTER_ID, PlayerId } from '../common/EntityIds';

import { PeriodicPieceGeneratorFactory } from './board-generation/createPeriodicPieceGenerator';
import { PeriodicPieceGenerator } from './board-generation/PeriodicPieceGenerator';

import { CommunicationRequestsStore } from './communication/CommunicationRequestsStore';
import { getGameStartedMessagePayload } from './communication/getGameStartedMessagePayload';

export class Game {
  public board: Board;
  public readonly playersContainer: PlayersContainer;
  public readonly definition: GameDefinition;
  public readonly scoreboard: Scoreboard;

  private readonly logger: LoggerInstance;
  private readonly periodicPieceGenerator: PeriodicPieceGenerator;
  private readonly playerMessageHandler: PlayerMessageHandler;
  private readonly communicator: Communicator;
  private readonly uiController: UIController;
  private readonly updateUI: Function;
  private readonly writeCsvLog: (message: Message<any>, playerId: PlayerId, valid: boolean) => any;
  private _state = GameState.Registered;

  public get state() {
    return this._state;
  }

  constructor(
    gameDefinition: GameDefinition,
    logger: LoggerInstance,
    uiController: UIController,
    communicator: Communicator,
    periodicPieceGeneratorFactory: PeriodicPieceGeneratorFactory,
    onPointsLimitReached: Function,
    updateUI: Function,
    writeCsvLog: (message: Message<any>, playerId: PlayerId, valid: boolean) => any
  ) {
    this.definition = gameDefinition;
    this.board = new Board(this.definition.boardSize, this.definition.goalLimit);
    this.scoreboard = new Scoreboard(this.definition.goalLimit);
    this.logger = logger;
    this.uiController = uiController;
    this.playersContainer = new PlayersContainer();
    this.communicator = communicator;
    this.updateUI = updateUI;
    this.writeCsvLog = writeCsvLog;
    this.periodicPieceGenerator = periodicPieceGeneratorFactory(this.board);

    this.playerMessageHandler = new PlayerMessageHandler(
      {
        board: this.board,
        playersContainer: this.playersContainer,
        actionDelays: this.definition.delays,
        logger: this.logger,
        scoreboard: this.scoreboard,
        sendMessage: this.sendIngameMessage.bind(this),
        onPointsLimitReached
      },
      new CommunicationRequestsStore()
    );
  }

  public start() {
    if (this.state === GameState.InProgress) {
      throw new Error('Game is already in progress');
    }

    this._state = GameState.InProgress;
    this.periodicPieceGenerator.init();
    this.sendGameStartedMessageToPlayers();
  }

  public stop() {
    if (this.state === GameState.InProgress) {
      this.periodicPieceGenerator.destroy();
    }

    this._state = GameState.Finished;
    this.sendGameFinishedMessageToPlayers();
  }

  public async handleMessage(message: Message<any>) {
    const messageSender = <Player>this.playersContainer.getPlayerById(message.senderId);

    const result = this.processPlayerMessage(message);
    if (!result.valid) {
      const actionInvalidMessage: ActionInvalidMessage = {
        type: 'ACTION_INVALID',
        recipientId: message.senderId,
        senderId: GAME_MASTER_ID,
        payload: {
          reason: result.reason
        }
      };

      if (messageSender) {
        this.writeCsvLog(
          message.type,
          message.senderId,
          messageSender.teamId,
          messageSender.isLeader,
          false
        );
      } else {
        this.logger.warn('Could not write csv log, invalid senderId');
      }

      return this.sendIngameMessage(actionInvalidMessage);
    }

    const actionValidMessage: ActionValidMessage = {
      type: 'ACTION_VALID',
      recipientId: message.senderId,
      senderId: GAME_MASTER_ID,
      payload: {
        delay: result.delay
      }
    };

    this.writeCsvLog(
      message.type,
      message.senderId,
      messageSender.teamId,
      messageSender.isLeader,
      true
    );

    this.updateUI();
    this.sendIngameMessage(actionValidMessage);
    this.sendIngameMessage(await result.responseMessage);
  }

  /**
   * This method is called internally by `handleMessage`.
   *
   * Left public only in order not to modify the tests by much.
   */
  public processPlayerMessage(message: Message<any>): ProcessMessageResult<any> {
    const sender = this.playersContainer.getPlayerById(message.senderId);
    if (!sender) {
      return {
        valid: false,
        reason: 'Sender ID is invalid'
      };
    }

    if (this.state !== GameState.InProgress) {
      return {
        valid: false,
        reason: 'Game is not in progress'
      };
    }

    if (sender.isBusy) {
      return {
        valid: false,
        reason: 'Sender is busy'
      };
    }

    sender.isBusy = true;

    const processMessageResult = this.playerMessageHandler.handleMessage(sender, message);
    if (processMessageResult.valid) {
      processMessageResult.responseMessage.then(() => (sender.isBusy = false));
    } else {
      sender.isBusy = false;
    }

    return processMessageResult;
  }

  public handlePlayerDisconnectedMessage(message: PlayerDisconnectedMessage) {
    const disconnectedPlayer = this.playersContainer.getPlayerById(message.payload.playerId);

    if (!disconnectedPlayer) {
      return;
    }

    this.removePlayer(disconnectedPlayer);
    this.uiController.updateBoard(this.board);
  }

  public removePlayer(disconnectedPlayer: Player) {
    this.board.removePlayer(disconnectedPlayer);
    this.playersContainer.removePlayer(disconnectedPlayer);
  }

  public addPlayer(player: Player) {
    this.playersContainer.addPlayer(player);
    this.board.addPlayer(player);
    this.updateUI();
  }

  public sendIngameMessage(message: Message<any>): void {
    if (this.state !== GameState.InProgress) {
      this.logger.verbose(
        `Message ${message.type} will not be sent because the game is not in progress`
      );

      return;
    }

    return this.communicator.sendMessage(message);
  }

  public async register() {
    const registerGameMessage: RegisterGameRequest = {
      type: 'REGISTER_GAME_REQUEST',
      senderId: GAME_MASTER_ID,
      payload: {
        game: this.definition
      }
    };

    this.logger.info('Registering the game');
    this.communicator.sendMessage(registerGameMessage);

    const response = <RegisterGameResponse>await this.communicator.waitForSpecificMessage(
      msg => msg.type === 'REGISTER_GAME_RESPONSE'
    );

    if (!response.payload.registered) {
      throw new Error('Cannot register the game');
    }
  }

  public async unregister() {
    const unregisterGameRequest: UnregisterGameRequest = {
      senderId: GAME_MASTER_ID,
      recipientId: COMMUNICATION_SERVER_ID,
      type: 'UNREGISTER_GAME_REQUEST',
      payload: {
        gameName: this.definition.name
      }
    };

    this.communicator.sendMessage(unregisterGameRequest);

    const response = <UnregisterGameResponse>await this.communicator.waitForSpecificMessage(
      msg => msg.type === 'UNREGISTER_GAME_RESPONSE'
    );

    if (!response.payload.unregistered) {
      throw new Error('Cannot unregister the game');
    }
  }

  public tryAcceptPlayer(message: PlayerHelloMessage) {
    const teamPlayers = this.playersContainer.getPlayersFromTeam(message.payload.teamId);

    if (this.state === GameState.InProgress) {
      throw new Error('Game already started');
    }

    if (teamPlayers.length >= this.definition.teamSizes[message.payload.teamId]) {
      throw new Error('Team is full');
    }

    const teamAlreadyHasLeader = teamPlayers.find(player => player.isLeader);
    if (message.payload.isLeader && teamAlreadyHasLeader) {
      throw new Error('Team already has a leader');
    }

    if (
      !message.payload.isLeader &&
      !teamAlreadyHasLeader &&
      teamPlayers.length + 1 === this.definition.teamSizes[message.payload.teamId]
    ) {
      throw new Error('Team does not have a leader');
    }

    const newPlayer = new Player();
    newPlayer.playerId = message.senderId;
    newPlayer.teamId = message.payload.teamId;
    newPlayer.isLeader = message.payload.isLeader;
    newPlayer.isBusy = false;

    this.addPlayer(newPlayer);
  }

  private sendGameStartedMessageToPlayers() {
    const gameStartedMessagePayload = getGameStartedMessagePayload(this.playersContainer);
    this.playersContainer.players.forEach(player => {
      const message: GameStartedMessage = {
        senderId: GAME_MASTER_ID,
        recipientId: player.playerId,
        type: 'GAME_STARTED',
        payload: gameStartedMessagePayload
      };

      this.communicator.sendMessage(message);
    });
  }

  private sendGameFinishedMessageToPlayers() {
    const gameFinishedMessagePayload: GameFinishedMessagePayload = {
      team1Score: this.scoreboard.team1Score,
      team2Score: this.scoreboard.team2Score
    };

    this.playersContainer.players.forEach(player => {
      const gameFinishedMessage: GameFinishedMessage = {
        type: 'GAME_FINISHED',
        senderId: GAME_MASTER_ID,
        recipientId: player.playerId,
        payload: gameFinishedMessagePayload
      };

      this.communicator.sendMessage(gameFinishedMessage);
    });
  }
}
