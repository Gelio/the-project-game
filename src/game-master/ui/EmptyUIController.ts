import { Service } from '../../interfaces/Service';

import { UIController as IUIController } from './IUIController';

import { LoggerFactory } from '../../common/logging/LoggerFactory';

export class EmptyUIController implements Service, IUIController {
  private readonly loggerFactory: LoggerFactory;

  constructor(loggerFactory: LoggerFactory) {
    this.loggerFactory = loggerFactory;
  }

  public init() {
    // Left empty intentionally
  }

  public destroy() {
    // Left empty intentionally
  }

  public render() {
    // Left empty intentionally
  }

  public updateBoard() {
    // Left empty intentionally
  }

  public updateGameInfo() {
    // Left empty intentionally
  }

  public createLogger() {
    return this.loggerFactory.createConsoleLogger();
  }
}
