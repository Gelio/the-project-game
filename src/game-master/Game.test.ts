import { LoggerInstance } from 'winston';

import { LoggerFactory } from '../common/logging/LoggerFactory';

import { ActionDelays } from '../interfaces/ActionDelays';
import { BoardSize } from '../interfaces/BoardSize';

import { DiscoveryRequest } from '../interfaces/requests/DiscoveryRequest';

import { Game } from './Game';
import { PlayersContainer } from './PlayersContainer';
import { InvalidMessageResult } from './ProcessMessageResult';

import { UIController } from './ui/UIController';

import { Player } from './Player';

function createMockUiController(): UIController {
  return <any>{
    updateBoard: jest.fn()
  };
}

describe('[GM] Game', () => {
  const boardSize: BoardSize = {
    x: 30,
    goalArea: 50,
    taskArea: 300
  };
  let game: Game;
  let uiController: UIController;
  let loggerInstance: LoggerInstance;

  const actionDelays: ActionDelays = {
    communicationAccept: 4000,
    communicationRequest: 4000,
    destroy: 4000,
    discover: 4000,
    move: 4000,
    pick: 4000,
    test: 4000,
    place: 4000
  };

  let player: Player;

  beforeAll(() => {
    const loggerFactory = new LoggerFactory();
    loggerInstance = loggerFactory.createEmptyLogger();

    uiController = createMockUiController();
  });

  beforeEach(() => {
    const playersContainter = new PlayersContainer();
    const pointsLimit = 5;
    game = new Game(
      boardSize,
      pointsLimit,
      loggerInstance,
      uiController,
      playersContainter,
      actionDelays
    );

    player = new Player();
    player.playerId = 'player';
    player.teamId = 1;
    player.isLeader = true;
    player.isBusy = false;
    player.isConnected = true;
  });

  describe('processMessage', () => {
    it('should reject message from non existing player', () => {
      const message: DiscoveryRequest = {
        senderId: 'player1',
        type: 'DISCOVERY_REQUEST',
        payload: undefined
      };
      const processedMessageResult = game.processMessage(message);
      expect(processedMessageResult.valid).toBe(false);

      const invalidResult = <InvalidMessageResult>processedMessageResult;
      expect(invalidResult).toBeDefined();
      expect(invalidResult.reason).toMatchSnapshot();
    });

    it('should reject message from busy player', () => {
      game.addPlayer(player);

      player.isBusy = true;

      const message: DiscoveryRequest = {
        senderId: player.playerId,
        type: 'DISCOVERY_REQUEST',
        payload: undefined
      };
      const processedMessageResult = game.processMessage(message);
      expect(processedMessageResult.valid).toBe(false);

      const invalidResult = <InvalidMessageResult>processedMessageResult;
      expect(invalidResult).toBeDefined();
      expect(invalidResult.reason).toMatchSnapshot();
    });

    it('should process valid request', () => {
      game.addPlayer(player);

      const message: DiscoveryRequest = {
        senderId: player.playerId,
        type: 'DISCOVERY_REQUEST',
        payload: undefined
      };
      const processedMessageResult = game.processMessage(message);
      expect(processedMessageResult.valid).toBe(true);
    });
  });

  it('should start the game', () => {
    game.start();
    expect(game.hasStarted).toBe(true);
  });

  it('should stop the game', () => {
    game.stop();
    expect(game.hasStarted).toBe(false);
  });

  describe('after resetting the game', () => {
    it('players should receive new positions', () => {
      game.addPlayer(player);

      const oldPosition = player.position;
      game.reset();

      expect(player.position).not.toBe(oldPosition);
    });
  });

  it('should add player to the game', () => {
    game.addPlayer(player);

    expect(game.playersContainer.getPlayerById(player.playerId)).toEqual(player);
  });

  it('should remove player from the game', () => {
    game.addPlayer(player);

    game.removePlayer(player);

    expect(game.playersContainer.getPlayerById(player.playerId)).toBeUndefined();
  });
});
