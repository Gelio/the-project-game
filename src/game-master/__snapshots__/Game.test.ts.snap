// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GM] Game processMessage should reject message from busy player 1`] = `"Sender is busy"`;

exports[`[GM] Game processMessage should reject message from non existing player 1`] = `"Sender ID is invalid"`;
